core/shared/src/main/scala/fs2/util/Concurrent.scala

```diff
@@ -0,0 +1,299 @@
+package fs2
+package util
+
+import cats.{ Eq, Traverse }
+import cats.implicits.{ catsSyntaxEither => _, _ }
+import cats.effect.{ Effect, IO }
+
+import fs2.internal.{Actor,LinkedMap}
+import ExecutionContexts._
+import java.util.concurrent.atomic.{AtomicBoolean,AtomicReference}
+
+import scala.concurrent.ExecutionContext
+
+/**
+ * Type class which describes effects that support concurrency.
+ *
+ * Instances of this type class are defined by providing an implementation of `ref`,
+ * which allocates a mutable memory cell that supports various asynchronous operations.
+ */
+trait Concurrent[F[_]] extends Effect[F] { self =>
+
+  /** Creates an asynchronous, concurrent mutable reference. */
+  def ref[A]: F[Concurrent.Ref[F,A]]
+
+  /** Creates an asynchronous, concurrent mutable reference, initialized to `a`. */
+  def refOf[A](a: A): F[Concurrent.Ref[F,A]] = flatMap(ref[A])(r => map(r.setPure(a))(_ => r))
+
+  /** Like `traverse` but each `F[B]` computed from an `A` is evaluated in parallel. */
+  def parallelTraverse[G[_],A,B](g: G[A])(f: A => F[B])(implicit G: Traverse[G]): F[G[B]] =
+    flatMap(G.traverse(g)(f andThen start)(self)) { gfb => G.sequence(gfb)(self) }
+
+  /** Like `sequence` but each `F[A]` is evaluated in parallel. */
+  def parallelSequence[G[_],A](v: G[F[A]])(implicit G: Traverse[G]): F[G[A]] =
+    parallelTraverse(v)(identity)
+
+  /**
+   * Begins asynchronous evaluation of `f` when the returned `F[F[A]]` is
+   * bound. The inner `F[A]` will block until the result is available.
+   */
+  def start[A](f: F[A]): F[F[A]] =
+    flatMap(ref[A]) { ref =>
+      map(ref.set(f)) { _ => ref.get }
+    }
+
+  /**
+    * Returns an effect that, when run, races evaluation of `fa` and `fb`,
+    * and returns the result of whichever completes first. The losing effect
+    * continues to execute in the background though its result will be sent
+    * nowhere.
+   */
+  def race[A, B](fa: F[A], fb: F[B]): F[Either[A, B]]
+
+  /**
+   * Like `unsafeRunSync` but execution is shifted to the execution context used by this `Concurrent` instance.
+   * This method returns immediately after submitting execution to the execution context.
+   */
+  def unsafeRunAsync[A](fa: F[A])(f: Either[Throwable, A] => IO[Unit]): Unit
```

Comment:
@djspiewak Note that we can't move it to a companion b/c the execution context is captured in the `Concurrent` *instance*. I'll leave this here for now and then open a new PR after this one is merged that makes `Concurrent` a utility instead of a full type class.